TTD/ReplayApi/TraceFallbacks/TraceFallbacks.cpp

```diff
@@ -0,0 +1,458 @@
+    if (cmdLine.ShowInstructionCount && cmdLine.InputFile.has_value())
+    {
+        std::cerr << "Error: -instructioncount option can only be used with trace files, not input files\n";
+        return {};
+    }
+    
+    return cmdLine;
+}
+
+int main(int argc, char *argv[])
+{
+    std::optional<CommandLine> commandLine = TryParseCommandLine(argc, argv);
+    if (!commandLine.has_value())
+    {
+        std::cerr << "Usage: TraceFallbacks [reporting options] <trace file> [-o <json file>]\n";
+        std::cerr << "       TraceFallbacks [reporting options] -i <json file>\n";
+        std::cerr << "Options:\n";
+        std::cerr << "       -i <input file>   - Read fallback data from json file instead of trace file\n";
+        std::cerr << "       -o <output file>  - Write all fallback data to json\n";
+        std::cerr << "Reporting Options:\n";
+        std::cerr << "       -instructioncount - Show instruction counts for each thread and the entire trace\n";
+        std::cerr << "       -normalize        - Normalize instructions\n";
+        std::cerr << "       -top N            - Show top N fallbacks (default: 20)\n";
+        std::cerr << "       -all              - Show all fallbacks (equivalent to -top max)\n";
+        return 1;
+    }
+
+    // Check if we're in input file mode
+    if (commandLine->InputFile.has_value())
+    {
+        // Read fallback stats from file and display
+        std::vector<FallbackInfo> fallbackInfo;
+        std::error_code ec = ReadFallbackStatsFile(commandLine->InputFile.value(), fallbackInfo);
+        if (ec)
+        {
+            std::cout << std::format("Failed to read fallback stats from {}: {}\n", 
+                commandLine->InputFile.value().string(), ec.message());
+            return -1;
+        }
+        
+        std::cout << std::format("Read {} fallback entries from {}\n\n", 
+            fallbackInfo.size(), commandLine->InputFile.value().string());
+        
+        // Display the analysis
+        PrintFallbackStatsToConsole(fallbackInfo, commandLine.value());
+        
+        // Optionally write to output file
+        if (commandLine->OutputFile.has_value())
+        {
+            PrintFallbackStatsToFile(fallbackInfo, commandLine->OutputFile.value());
+        }
+        
+        return 0;
+    }
+
+    // Original trace file processing mode
+    std::filesystem::path traceFile = commandLine->TraceFile.value();
+
+    // Start timing the operation
+    auto const startTime = std::chrono::high_resolution_clock::now();
+    
+    // To get started, we need a replay engine to load the trace file
+    auto [pOwnedReplayEngine, createResult] = MakeReplayEngine();
+    if (createResult != 0 || pOwnedReplayEngine == nullptr)
+    {
+        std::cout << std::format("There was an issue creating a replay engine ({})\n", createResult);
+        return -1;
+    }
+
+    // This allows the tool to get any messages from the initialization of the engine
```

Comment:
Correction: gets messages throughout the engine's operation. Including any derailments or file corruption found.